pallets/vanchor/src/lib.rs

```diff
@@ -0,0 +1,353 @@
+
+		#[pallet::constant]
+		type PalletId: Get<PalletId>;
+
+		/// The tree type
+		type LinkableTree: LinkableTreeInterface<pallet_linkable_tree::LinkableTreeConfigration<Self, I>>
+			+ LinkableTreeInspector<pallet_linkable_tree::LinkableTreeConfigration<Self, I>>;
+
+		/// The verifier
+		type Verifier: VerifierModule;
+
+		/// Currency type for taking deposits
+		type Currency: MultiCurrency<Self::AccountId>;
+
+		type PostDepositHook: PostDepositHook<Self, I>;
+
+		/// Native currency id
+		#[pallet::constant]
+		type NativeCurrencyId: Get<CurrencyIdOf<Self, I>>;
+
+		/// Weight info for pallet
+		type WeightInfo: WeightInfo;
+	}
+
+	/// The map of trees to their anchor metadata
+	#[pallet::storage]
+	#[pallet::getter(fn anchors)]
+	pub type Anchors<T: Config<I>, I: 'static = ()> = StorageMap<
+		_,
+		Blake2_128Concat,
+		T::TreeId,
+		Option<AnchorMetadata<T::AccountId, BalanceOf<T, I>, CurrencyIdOf<T, I>>>,
+		ValueQuery,
+	>;
+
+	/// The map of trees to their spent nullifier hashes
+	#[pallet::storage]
+	#[pallet::getter(fn nullifier_hashes)]
+	pub type NullifierHashes<T: Config<I>, I: 'static = ()> =
+		StorageDoubleMap<_, Blake2_128Concat, T::TreeId, Blake2_128Concat, T::Element, bool, ValueQuery>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config<I>, I: 'static = ()> {
+		/// New tree created
+		AnchorCreation { tree_id: T::TreeId },
+		/// Amount has been withdrawn from the anchor
+		Withdraw { who: T::AccountId, amount: BalanceOf<T, I> },
+		/// Amount has been deposited into the anchor
+		Deposit {
+			depositor: T::AccountId,
+			tree_id: T::TreeId,
+			leaf: T::Element,
+			amount: BalanceOf<T, I>,
+		},
+		/// Post deposit hook has executed successfully
+		PostDeposit {
+			depositor: T::AccountId,
+			tree_id: T::TreeId,
+			leaf: T::Element,
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T, I = ()> {
+		/// Invalid Merkle Roots
+		InvalidMerkleRoots,
+		/// Unknown root
+		UnknownRoot,
+		/// Invalid withdraw proof
+		InvalidWithdrawProof,
+		/// Mixer not found.
+		NoAnchorFound,
+		/// Invalid nullifier that is already used
+		/// (this error is returned when a nullifier is used twice)
+		AlreadyRevealedNullifier,
+	}
+
+	#[pallet::hooks]
+	impl<T: Config<I>, I: 'static> Hooks<BlockNumberFor<T>> for Pallet<T, I> {}
+
+	#[pallet::call]
+	impl<T: Config<I>, I: 'static> Pallet<T, I> {
+		#[pallet::weight(<T as Config<I>>::WeightInfo::create(*depth as u32, *max_edges))]
+		pub fn create(
+			origin: OriginFor<T>,
+			max_edges: u32,
+			depth: u8,
+			asset: CurrencyIdOf<T, I>,
+		) -> DispatchResultWithPostInfo {
+			// Should it only be the root who can create anchors?
+			ensure_root(origin)?;
+			let tree_id = <Self as VAnchorInterface<_>>::create(T::AccountId::default(), depth, max_edges, asset)?;
+			Self::deposit_event(Event::AnchorCreation { tree_id });
+			Ok(().into())
+		}
+
+		#[pallet::weight(0)] // TODO: Fix after benchmarks
+		pub fn transact(
```

Comment:
Consider using structs similar to the [solidity](https://github.com/webb-tools/protocol-solidity/blob/main/contracts/vanchors/VAnchorBase.sol#L130):
- https://github.com/webb-tools/protocol-solidity/blob/main/contracts/vanchors/VAnchorBase.sol#L69
- https://github.com/webb-tools/protocol-solidity/blob/main/contracts/vanchors/VAnchorBase.sol#L78